crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,529 @@
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if !batch.is_empty() {
+                    let results: Vec<_> = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| NativePriceResult {
+                                token,
+                                result: price,
+                            })
+                            .collect(),
+                        Err(err) => {
+                            tracing::error!(?err, "failed to send native price batch request");
+                            batch
+                                .into_iter()
+                                .map(|token| NativePriceResult {
+                                    token,
+                                    result: Err(err.clone()),
+                                })
+                                .collect()
+                        }
+                    };
+                    for result in results {
+                        let _ = broadcast_sender.send(result);
+                    }
+                }
+            }
+        }))
+    }
+
+    /// Request to get estimate prices in a batch
+    pub async fn request_buffered_estimate_prices(
+        &self,
+        token: &H160,
+    ) -> NativePriceEstimateResult {
+        // Sends the token for requesting price
+        self.calls.unbounded_send(*token).map_err(|e| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "failed to append a new token to the queue: {e:?}"
+            ))
+        })?;
+
+        let mut rx = self.broadcast_sender.subscribe();
+
+        tokio::time::timeout(self.config.result_ready_timeout, async {
+            loop {
+                if let Ok(Some(result)) =
+                    Self::receive_with_timeout(&mut rx, token, self.config.batch_delay).await
+                {
+                    return result.result;
+                }
+            }
+        })
+        .await
+        .map_err(|_| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "blocking buffered estimate prices timeout elapsed"
+            ))
+        })?
+    }
+
+    // Function to receive with a timeout
+    async fn receive_with_timeout(
+        rx: &mut broadcast::Receiver<NativePriceResult>,
+        token: &H160,
+        timeout_duration: Duration,
+    ) -> Result<Option<NativePriceResult>, Elapsed> {
+        tokio::time::timeout(timeout_duration, async {
+            match rx.recv().await {
+                Ok(value) => (value.token == *token).then_some(value),
+                Err(_) => None,
```

Comment:
Just to pin point one of the things brought up during the discussion above.
This case only means the receiver missed at least 1 one item in the stream. It can very well be that the item it is interested in is still in the stream.
So I suggest in the `None` case to exhaust the channel with a `try_recv` loop hoping to still find the desired item.

Reply:
Good point! But the requested value may not necessarily be ready yet, so if we do `try_recv()` may yields only errors until the token price is generated and sent to the broadcast. As it is right now, if it is lagged behind, it just keeps consuming elements until the desired one is found or generated, The only thing is that the operation is `async`.

Reply:
On the other hand the `RecvError::Closed` could be better handled :thinking:

Reply:
Oh, right. Overlooked that you are already looping somewhere else. In that case I think this is fine.